src/ccache.rs

```diff
@@ -0,0 +1,445 @@
+}
+
+#[binwrite]
+#[bw(big)]
+struct Addresses {
+    #[bw(try_calc(u32::try_from(addresses.len())))]
+    count: u32,
+    addresses: Vec<Address>,
+}
+
+#[binwrite]
+#[bw(big)]
+struct AuthDataComponent {
+    ad_type: u16,
+    data: DataComponent,
+}
+
+#[binwrite]
+#[bw(big)]
+struct AuthData {
+    #[bw(try_calc(u32::try_from(auth_data.len())))]
+    count: u32,
+    auth_data: Vec<AuthDataComponent>,
+}
+
+#[binwrite]
+#[bw(big)]
+struct CredentialV4 {
+    client: PrincipalV4,
+    server: PrincipalV4,
+    keyblock: KeyBlock,
+    authtime: u32,
+    starttime: u32,
+    endtime: u32,
+    renew_till: u32,
+    is_skey: u8,
+    ticket_flags: u32,
+    addresses: Addresses,
+    authdata: AuthData,
+    ticket: DataComponent,
+    second_ticket: DataComponent,
+}
+
+#[binwrite]
+#[bw(big)]
+enum Credential {
+    V4(CredentialV4),
+}
+
+#[binwrite]
+#[bw(big, magic = 4u8)]
+pub(crate) struct FileCredentialCacheV4 {
+    header: FileCredentialCacheHeader,
+    principal: FileCredentialCachePrincipal,
+    credentials: Vec<Credential>,
+}
+
+#[binwrite]
+#[bw(big, magic = 5u8)]
+pub(crate) enum FileCredentialCache {
+    V4(FileCredentialCacheV4),
+}
+
+#[binwrite]
+#[bw(big)]
+pub(crate) enum CredentialCache {
+    File(FileCredentialCache),
+}
+
+impl TryFrom<&Name> for PrincipalV4 {
+    type Error = KrbError;
+
+    fn try_from(name: &Name) -> Result<Self, Self::Error> {
+        match name {
+            Name::Principal { name, realm } => {
+                let p: PrincipalV4 = PrincipalV4 {
+                    name_type: 1 as u32,
+                    realm: DataComponent {
+                        value: realm.as_bytes().into(),
+                    },
+                    components: vec![DataComponent {
+                        value: name.as_bytes().into(),
+                    }],
+                };
+                Ok(p)
+            }
+            Name::SrvInst { service, realm } => {
+                let p: PrincipalV4 = PrincipalV4 {
+                    name_type: 2 as u32,
+                    realm: DataComponent {
+                        value: realm.as_bytes().into(),
+                    },
+                    components: vec![
+                        DataComponent {
+                            value: service.as_bytes().into(),
+                        },
+                        DataComponent {
+                            value: realm.as_bytes().into(),
+                        },
+                    ],
+                };
+                Ok(p)
+            }
+            _ => todo!(),
+        }
+    }
+}
+
+impl TryFrom<&SessionKey> for KeyBlockV4 {
+    type Error = KrbError;
+
+    fn try_from(value: &SessionKey) -> Result<Self, Self::Error> {
+        match value {
+            SessionKey::Aes256CtsHmacSha196 { k } => Ok(KeyBlockV4 {
+                enc_type: 0x12,
+                data: DataComponent { value: k.to_vec() },
+            }),
+        }
+    }
+}
+
+impl FileCredentialCache {
```

Comment:
You can actually move this impl into a mod test {} since impls can be outside the same module from where the struct is. That might help you need less cfg[test] sprinkled around :)

Reply:
Done!